### PR TITLE
Presets! (and more)

### PR DESCRIPTION
**Pull request by topisani**
There's a lot going on, sorry it all ended up here

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -166,6 +166,7 @@ install:
 before_script:
   - cd "${TRAVIS_BUILD_DIR}"
   - mkdir build && cd build
+  - export CXXFLAGS="$CXXFLAGS -Werror"
```
Comment (dacarley) [inline .travis.yml]: Why not leave this turned on CMakeLists.txt?

File: external/CMakeLists.txt [hunk 2]
```diff
@@ -16,18 +16,18 @@ file(GLOB_RECURSE external_src
   "${CMAKE_CURRENT_SOURCE_DIR}/src/*.cpp"
   )
 add_library(external ${external_src})
-target_include_directories(external PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
+target_include_directories(external SYSTEM PUBLIC  ${CMAKE_CURRENT_SOURCE_DIR}/include)
```
Comment (dacarley) [inline external/CMakeLists.txt]: What does `SYSTEM` do?

File: src/core/engines/engine_dispatcher.cpp [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
 
-  template<EngineType ET>
-  const Engine<ET>& EngineDispatcher<ET>::current() const noexcept
-  {
-    return *_current;
-  }
+// EngineDispatcher Implementations /////////////////////////////////////////
+template<EngineType ET>
+void EngineDispatcher<ET>::init()
+{
```
Comment (dacarley) [inline src/core/engines/engine_dispatcher.cpp]: Is the indentation in this file as screwy as it looks in Github?

File: external/CMakeLists.txt (hunk 2 above)
Comment (topisani) [reply external/CMakeLists.txt]: makes it a system header so we dont get warnings from it

File: src/core/engines/engine_dispatcher.cpp (hunk 3 above)
Comment (topisani) [reply src/core/engines/engine_dispatcher.cpp]: woops yeah

File: .travis.yml (hunk 1 above)
Comment (topisani) [reply .travis.yml]: emacs gave me warnings as errors, and i thought in general its best to just have it on CI

**Comment by topisani** [comment]
Not quite done yet, but side effects have so far been fixing the `recursive_directory_iterator` and loguru.

I got a segfault while playing around with the menu, not sure about that, probably a lack of bounds checking somewhere.

Im thinking the engines should store the currently loaded preset (even if it has been modified) so you can allways see which one is in use.

Theres also some thoughts to be had about the relationship with state in general. Im thinking the preset is loaded first, and then the options in `state.json` are loaded on top. This way we dont need a default `state.json`, since it will just load the first preset by default.

Another thing is the debug UI. 
I think a menu to create a preset with the current settings would be very helpful when creating presets, and i'll try to get that in with this PR too

**Comment by topisani** [comment]
Alright, lets see

The engines now serialize the preset they currently have loaded, and startup with the first preset (if avaliable) by default.

**Comment by topisani** [comment]
Alright, lets see

The engines now serialize the preset they currently have loaded, and startup with the first preset (if avaliable) by default. the stored state is loaded on top.

presets have been moved into their own service.

the debug ui has a window for presets, which can be used to create a preset for the current engine with the current settings.

**Pull request merged**